group/views.py

```diff
@@ -102,41 +95,58 @@ def get_all_announce(request, group_id):
 
 def detail(request, group_id):
 
+    redirect_id = 1; #return to detail page.
     group = get_group(group_id)
     show_number = 5; #number for brief list to show in group detail page.
-    all_contest = group.trace_contest.all()
+    all_contest = group.trace_contest.order_by('-start_time')
     annowence_list = group.announce.all()
     student_list = group.member.order_by('user_level')
-    coowner_list = group.coowner.all()
     owner = group.owner
-    user_is_owner = has_group_ownership(request.user, group)
+    form = AnnounceForm()
+
+    user = validate_user(request.user)
+    user_is_owner = has_group_ownership(user, group)
+    user_is_coowner = has_group_coownership(user, group)
+
+    user_has_auth = user_is_owner or user_is_coowner
 
     running_contest_list = []
     ended_contest_list = []
     now = timezone.now()
-    for contest in all_contest:
-        if contest.start_time < now and contest.end_time > now:
-            running_contest_list.append(contest)
-        elif contest.end_time < now:
-            ended_contest_list.append(contest)
+    running_contest_list = group.trace_contest.filter(start_time__lte=now, end_time__gte=now)
```

Comment:
It's better to use ```running_contest_list = group.trace_contest.filter(start_time__lte=now, end_time__gte=now)[:show_number]```
Limited query has better performance than query all objects then get some of them in later code.
In this way, line 123 can be modified to ```'running_contest_list': running_contest_list```